yellowstone-grpc-tools/src/scylladb/sink.rs

```diff
@@ -476,36 +617,62 @@ impl ScyllaSink {
 
         let session = Arc::new(session);
 
-        let shard_count = SHARD_COUNT;
+        let producer_info = get_producer_info_by_id(Arc::clone(&session), producer_id)
+            .await?
+            .unwrap_or_else(|| panic!("producer {:?} has not yet been registered", producer_id));
+
+        let producer_lock =
+            try_acquire_lock(Arc::clone(&session), producer_id, config.ifname.to_owned()).await?;
+        let shard_count = producer_info.num_shards as usize;
+
+        init_producer_commit_log(Arc::clone(&session), producer_id, shard_count).await?;
 
         let mut sharders = vec![];
 
         info!("Will create {:?} shards", shard_count);
-        let maybe_shard_offsets =
-            get_shard_offsets_for_producer(Arc::clone(&session), producer_id).await?;
-        let shard_offsets = maybe_shard_offsets.unwrap_or(vec![1; shard_count]);
-        for (shard_id, next_offset) in shard_offsets.iter().enumerate() {
+        let shard_offsets =
+            get_max_shard_offsets_for_producer(Arc::clone(&session), producer_id, shard_count)
+                .await?;
+
+        let mut shard_handles = Vec::with_capacity(shard_count);
+        for (shard_id, last_offset) in shard_offsets.into_iter() {
             let session = Arc::clone(&session);
             let shard = Shard::new(
                 session,
-                shard_id as i16,
+                shard_id,
                 producer_id,
-                *next_offset,
+                last_offset + 1,
                 DEFAULT_SHARD_MAX_BUFFER_CAPACITY,
                 config.batch_size_kb_limit * 1024,
                 config.linger,
             );
-            let shard_mailbox = shard.into_daemon();
+            let (shard_mailbox, shard_handle) = shard.into_daemon();
+            shard_handles.push(shard_handle);
             sharders.push(shard_mailbox);
         }
 
         let sender = spawn_round_robin(sharders);
 
         Ok(ScyllaSink {
             router_handle: sender,
+            shard_handles,
+            producer_lock,
         })
     }
 
+    pub async fn shutdown(self) -> anyhow::Result<()> {
+        warn!("Shutthing down scylla sink...");
+        self.router_handle.send(ClientCommand::Shutdown).await?;
+        for (i, ss) in self.shard_handles.into_iter().enumerate() {
+            let result = ss.await;
+            if result.is_err() {
+                error!("shard {} error: {:?}", i, result);
+            }
```

Comment:
```suggestion
            if let Err(error) = ss.await {
                error!("shard {i} error: {error:?}");
            }
```
should be more readable than `shard 0 error: Err(error_here)` (we do not need `Err()` :thinking: )